### PR TITLE
Adding Infoblox Stix Shifter Module

### PR DESCRIPTION
**Pull request by grimmjow8**
Infoblox connector implementation. 

Support for 3 dialects (apis)
- TIDE DB
- DNS Events
- Dossier

developed with https://github.com/quinoxe

### CODE REVIEW

**Comment by grimmjow8** [comment]
FYI resolved the conflict with the unittests for the new Infoblox module. Ran the tests and all passed except for those in the Onelogin module. Verified the same tests are failing on the main `develop` branch.

**Review by mdazam1942 — COMMENTED** [review]
I see a strange behavior for `ipv4-addr` pattern translation if Combined Comparisons and Observation operators is used. Not sure if I am missing something. Is there any particular reason? Here's an example of the error:

```
% python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"  

{
    "success": false,
    "code": "not_implemented",
    "error": "wrong parameter : Multiple criteria for one field is not support in Infoblox connector"
}
```

File: stix_shifter_modules/infoblox/stix_transmission/api_client.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+                                    )
+
+    def ping_data_source(self):
+        # Pings the data source
+        endpoint = 'tide/api/data/threats/state'
+        # now = datetime.datetime.utcnow().isoformat(timespec="milliseconds") + "Z"
+        # https://csp.infoblox.com:443/tide/api/data/threats/state?type=host&rlimit=1
+
+        resp = self.client.call_api(endpoint, 'GET', timeout=self.timeout, urldata={"type": "host", "rlimit": "1"})
+        return {"code": resp.code}
```
Comment (mdazam1942) [inline stix_shifter_modules/infoblox/stix_transmission/api_client.py:44]: I would return the whole response instead of just the error code to have a better explanation of the error.

`return self.client.call_api(endpoint, 'GET', timeout=self.timeout, urldata={"type": "host", "rlimit": "1"})`

**Comment by grimmjow8** [comment]
> I see a strange behavior for ipv4-addr pattern translation if Combined Comparisons and Observation operators is used. Not sure if I am missing something. Is there any particular reason? Here's an example of the error:
> python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"

Thanks @mdazam1942 for reviewing this! That is expected behavior. Currently, none of the Infoblox apis support searching in a manner that would support this. Looking at several of the other connectors, they produce queries like:
```
(sourceAddress = 127.0.0.2 OR destinationAddress = 127.0.0.2) AND (sourceAddress = 127.0.0.1 OR destinationAddress = 127.0.0.1)
```

None of the infoblox apis support `OR` and they all have a single `ip` type field to query, rather than separate source/destination. For Dossier there is only `IP`, for DNS Event there is only `qip`, and for TIDE DB there is only `ip`. So we ended up following what `trendmicro_vision_one` does and raise an error when multiple `ipv4-addr` are present.

File: stix_shifter_modules/infoblox/stix_transmission/api_client.py (hunk 1 above)
Comment (grimmjow8) [reply stix_shifter_modules/infoblox/stix_transmission/api_client.py]: Updated, here is an example error response:
```
❯ python3 main.py transmit infoblox '{"host":"csp.infoblox.com"}' '{"auth": {"token":"xxx"} }' ping
{
    "success": false,
    "error": "Invalid type --- type [host, ip, url, hash, email] must be specified in path or parameters\n",
    "code": "invalid_parameter"
}
```

**Comment by mdazam1942** [comment]
> python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"

I understand. Multiple IP address searches are very common. So I wonder if there's any workaround. The API not necessarily need support for source and destination IPs. Does the API accept searches if you add multiple `ip/qip/ip` in the query string like `&ip=127.0.0.1&ip=127.0.0.2`? you can also break down the pattern into multiple queries. Each query for one comparison expression.

**Comment by grimmjow8** [comment]
> > python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"

@mdazam1942 I've reworked the pattern parsing in the connector to be able to support AND/OR. In the ipv4-addr example, you'll now find it generating two queries per dialect (instead of throwing an error).

```
❯ python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"


 2021-08-29 16:30:16,310 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
 2021-08-29 16:30:16,313 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
 2021-08-29 16:30:16,314 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
{
    "queries": [
        "{\"offset\": 0, \"query\": \"from_date=2021-06-01T08:43:10.000Z&to_date=2021-08-31T10:43:10.000Z&ip=127.0.0.2\", \"threat_type\": \"ip\", \"source\": \"tideDbData\"}",
        "{\"offset\": 0, \"query\": \"from_date=2021-06-01T08:43:10.000Z&to_date=2021-08-31T10:43:10.000Z&ip=127.0.0.1\", \"threat_type\": \"ip\", \"source\": \"tideDbData\"}",
        "{\"offset\": 0, \"query\": \"t0=1622536990&t1=1630406590&qip=127.0.0.2\", \"threat_type\": null, \"source\": \"dnsEventData\"}",
        "{\"offset\": 0, \"query\": \"t0=1622536990&t1=1630406590&qip=127.0.0.1\", \"threat_type\": null, \"source\": \"dnsEventData\"}",
        "{\"offset\": 0, \"query\": \"value=127.0.0.2\", \"threat_type\": \"ip\", \"source\": \"dossierData\"}",
        "{\"offset\": 0, \"query\": \"value=127.0.0.1\", \"threat_type\": \"ip\", \"source\": \"dossierData\"}"
    ]
}

```

**Comment by mdazam1942** [comment]
> > > python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"
> 
> @mdazam1942 I've reworked the pattern parsing in the connector to be able to support AND/OR. In the ipv4-addr example, you'll now find it generating two queries per dialect (instead of throwing an error).
> 
> ```
> ❯ python main.py translate infoblox query '{}' "[ipv4-addr:value = '127.0.0.1' AND ipv4-addr:value = '127.0.0.2'] START t'2021-06-01T08:43:10Z' STOP t'2021-08-31T10:43:10Z'"
> 
> 
>  2021-08-29 16:30:16,310 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
>  2021-08-29 16:30:16,313 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
>  2021-08-29 16:30:16,314 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
> {
>     "queries": [
>         "{\"offset\": 0, \"query\": \"from_date=2021-06-01T08:43:10.000Z&to_date=2021-08-31T10:43:10.000Z&ip=127.0.0.2\", \"threat_type\": \"ip\", \"source\": \"tideDbData\"}",
>         "{\"offset\": 0, \"query\": \"from_date=2021-06-01T08:43:10.000Z&to_date=2021-08-31T10:43:10.000Z&ip=127.0.0.1\", \"threat_type\": \"ip\", \"source\": \"tideDbData\"}",
>         "{\"offset\": 0, \"query\": \"t0=1622536990&t1=1630406590&qip=127.0.0.2\", \"threat_type\": null, \"source\": \"dnsEventData\"}",
>         "{\"offset\": 0, \"query\": \"t0=1622536990&t1=1630406590&qip=127.0.0.1\", \"threat_type\": null, \"source\": \"dnsEventData\"}",
>         "{\"offset\": 0, \"query\": \"value=127.0.0.2\", \"threat_type\": \"ip\", \"source\": \"dossierData\"}",
>         "{\"offset\": 0, \"query\": \"value=127.0.0.1\", \"threat_type\": \"ip\", \"source\": \"dossierData\"}"
>     ]
> }
> ```
> 
> It still generates single queries where applicable to hopefully match the intent of the provided pattern. In the below example, `x-infoblox-threat:host_name` and `x-infoblox-threat:profile` only exist within TIDE DB, so only the query is generated and returned.
> 
> ```
> ❯ python main.py translate infoblox query '{}' "[x-infoblox-threat:host_name = 'example.host.com' AND domain-name:value = 'example.domain.com' AND x-infoblox-threat:profile = 'profile1']"
> 
> 
>  2021-08-29 16:35:43,311 stix_shifter_modules.infoblox.stix_translation.query_translator INFO     Converting STIX2 Pattern to data source query
> {
>     "queries": [
>         "{\"offset\": 0, \"query\": \"period=5 minutes&profile=profile1&domain=example.domain.com&host=example.host.com\", \"threat_type\": \"host\", \"source\": \"tideDbData\"}"
>     ]
> }
> ```

Great.looks good. this would be consistent with other connectors.

**Review by mdazam1942 — COMMENTED** [review]
@grimmjow8 Everything looks ok. But somehow the unittest coverage report is showing that coverage decreased. Can you have a look and see if it can be adjusted?

**Comment by grimmjow8** [comment]
> @grimmjow8 Everything looks ok. But somehow the unittest coverage report is showing that coverage decreased. Can you have a look and see if it can be adjusted?

@mdazam1942 Does the code coverage ignore test unittest classes? Looking over the report most of the classes are at 80+%. The major problems seemed to come from the test classes that have a lot of data resulting in a large number of uncovered lines.

I've does a few things in the hopes of increasing coverage:
- for a few classes, implemented some tests for the harder to cover test cases
- in test class, moved a large block of data to a .json file
- in test class, collapsed multi-line dicts to a few lines as possible

Running coverage locally, these changes seemed to provide a rather large bump. If this does not resolve it, there are a few other edge cases I might be able to write a test for.
 
Before:
```
----------- coverage: platform linux, python 3.8.5-final-0 -----------
Name                                                                        Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------
stix_shifter_modules/infoblox/__init__.py                                       0      0   100%
stix_shifter_modules/infoblox/entry_point.py                                   40      0   100%
stix_shifter_modules/infoblox/stix_translation/__init__.py                      0      0   100%
stix_shifter_modules/infoblox/stix_translation/query_constructor.py           268     89    67%
stix_shifter_modules/infoblox/stix_translation/query_translator.py             10      2    80%
stix_shifter_modules/infoblox/stix_translation/results_translator.py           19      4    79%
stix_shifter_modules/infoblox/stix_translation/transformers.py                 26      5    81%
stix_shifter_modules/infoblox/stix_transmission/__init__.py                     0      0   100%
stix_shifter_modules/infoblox/stix_transmission/api_client.py                 151     79    48%
stix_shifter_modules/infoblox/stix_transmission/delete_connector.py             8      0   100%
stix_shifter_modules/infoblox/stix_transmission/error_mapper.py                20      2    90%
stix_shifter_modules/infoblox/stix_transmission/ping_connector.py              21      0   100%
stix_shifter_modules/infoblox/stix_transmission/query_connector.py              6      0   100%
stix_shifter_modules/infoblox/stix_transmission/results_connector.py           22      0   100%
stix_shifter_modules/infoblox/stix_transmission/status_connector.py             4      0   100%
stix_shifter_modules/infoblox/test/stix_translation/test_json_to_stix.py      352    234    34%
stix_shifter_modules/infoblox/test/stix_translation/test_stix_to_query.py    1023   1000     2%
stix_shifter_modules/infoblox/test/stix_transmission/test_transmission.py     520    254    51%
-----------------------------------------------------------------------------------------------
TOTAL                                                                        2490   1669    33%

```

After:
```
----------- coverage: platform linux, python 3.8.5-final-0 -----------
Name                                                                        Stmts   Miss  Cover
-----------------------------------------------------------------------------------------------
stix_shifter_modules/infoblox/__init__.py                                       0      0   100%
stix_shifter_modules/infoblox/entry_point.py                                   40      0   100%
stix_shifter_modules/infoblox/stix_translation/__init__.py                      0      0   100%
stix_shifter_modules/infoblox/stix_translation/query_constructor.py           268     86    68%
stix_shifter_modules/infoblox/stix_translation/query_translator.py             10      2    80%
stix_shifter_modules/infoblox/stix_translation/results_translator.py           19      4    79%
stix_shifter_modules/infoblox/stix_translation/transformers.py                 26      5    81%
stix_shifter_modules/infoblox/stix_transmission/__init__.py                     0      0   100%
stix_shifter_modules/infoblox/stix_transmission/api_client.py                 141      2    99%
stix_shifter_modules/infoblox/stix_transmission/delete_connector.py             8      0   100%
stix_shifter_modules/infoblox/stix_transmission/error_mapper.py                20      2    90%
stix_shifter_modules/infoblox/stix_transmission/ping_connector.py              21      0   100%
stix_shifter_modules/infoblox/stix_transmission/query_connector.py              6      0   100%
stix_shifter_modules/infoblox/stix_transmission/results_connector.py           22      0   100%
stix_shifter_modules/infoblox/stix_transmission/status_connector.py             4      0   100%
stix_shifter_modules/infoblox/test/stix_translation/test_json_to_stix.py      208     20    90%
stix_shifter_modules/infoblox/test/stix_translation/test_stix_to_query.py      62      8    87%
stix_shifter_modules/infoblox/test/stix_transmission/test_transmission.py     422     52    88%
-----------------------------------------------------------------------------------------------
TOTAL                                                                        1277    181    86%

```

**Comment by mdazam1942** [comment]
> > @grimmjow8 Everything looks ok. But somehow the unittest coverage report is showing that coverage decreased. Can you have a look and see if it can be adjusted?
> 
> @mdazam1942 Does the code coverage ignore test unittest classes? Looking over the report most of the classes are at 80+%. The major problems seemed to come from the test classes that have a lot of data resulting in a large number of uncovered lines.
> 
> I've does a few things in the hopes of increasing coverage:
> 
> * for a few classes, implemented some tests for the harder to cover test cases
> * in test class, moved a large block of data to a .json file
> * in test class, collapsed multi-line dicts to a few lines as possible
> 
> Running coverage locally, these changes seemed to provide a rather large bump. If this does not resolve it, there are a few other edge cases I might be able to write a test for.
> 
> Before:
> 
> ```
> ----------- coverage: platform linux, python 3.8.5-final-0 -----------
> Name                                                                        Stmts   Miss  Cover
> -----------------------------------------------------------------------------------------------
> stix_shifter_modules/infoblox/__init__.py                                       0      0   100%
> stix_shifter_modules/infoblox/entry_point.py                                   40      0   100%
> stix_shifter_modules/infoblox/stix_translation/__init__.py                      0      0   100%
> stix_shifter_modules/infoblox/stix_translation/query_constructor.py           268     89    67%
> stix_shifter_modules/infoblox/stix_translation/query_translator.py             10      2    80%
> stix_shifter_modules/infoblox/stix_translation/results_translator.py           19      4    79%
> stix_shifter_modules/infoblox/stix_translation/transformers.py                 26      5    81%
> stix_shifter_modules/infoblox/stix_transmission/__init__.py                     0      0   100%
> stix_shifter_modules/infoblox/stix_transmission/api_client.py                 151     79    48%
> stix_shifter_modules/infoblox/stix_transmission/delete_connector.py             8      0   100%
> stix_shifter_modules/infoblox/stix_transmission/error_mapper.py                20      2    90%
> stix_shifter_modules/infoblox/stix_transmission/ping_connector.py              21      0   100%
> stix_shifter_modules/infoblox/stix_transmission/query_connector.py              6      0   100%
> stix_shifter_modules/infoblox/stix_transmission/results_connector.py           22      0   100%
> stix_shifter_modules/infoblox/stix_transmission/status_connector.py             4      0   100%
> stix_shifter_modules/infoblox/test/stix_translation/test_json_to_stix.py      352    234    34%
> stix_shifter_modules/infoblox/test/stix_translation/test_stix_to_query.py    1023   1000     2%
> stix_shifter_modules/infoblox/test/stix_transmission/test_transmission.py     520    254    51%
> -----------------------------------------------------------------------------------------------
> TOTAL                                                                        2490   1669    33%
> ```
> 
> After:
> 
> ```
> ----------- coverage: platform linux, python 3.8.5-final-0 -----------
> Name                                                                        Stmts   Miss  Cover
> -----------------------------------------------------------------------------------------------
> stix_shifter_modules/infoblox/__init__.py                                       0      0   100%
> stix_shifter_modules/infoblox/entry_point.py                                   40      0   100%
> stix_shifter_modules/infoblox/stix_translation/__init__.py                      0      0   100%
> stix_shifter_modules/infoblox/stix_translation/query_constructor.py           268     86    68%
> stix_shifter_modules/infoblox/stix_translation/query_translator.py             10      2    80%
> stix_shifter_modules/infoblox/stix_translation/results_translator.py           19      4    79%
> stix_shifter_modules/infoblox/stix_translation/transformers.py                 26      5    81%
> stix_shifter_modules/infoblox/stix_transmission/__init__.py                     0      0   100%
> stix_shifter_modules/infoblox/stix_transmission/api_client.py                 141      2    99%
> stix_shifter_modules/infoblox/stix_transmission/delete_connector.py             8      0   100%
> stix_shifter_modules/infoblox/stix_transmission/error_mapper.py                20      2    90%
> stix_shifter_modules/infoblox/stix_transmission/ping_connector.py              21      0   100%
> stix_shifter_modules/infoblox/stix_transmission/query_connector.py              6      0   100%
> stix_shifter_modules/infoblox/stix_transmission/results_connector.py           22      0   100%
> stix_shifter_modules/infoblox/stix_transmission/status_connector.py             4      0   100%
> stix_shifter_modules/infoblox/test/stix_translation/test_json_to_stix.py      208     20    90%
> stix_shifter_modules/infoblox/test/stix_translation/test_stix_to_query.py      62      8    87%
> stix_shifter_modules/infoblox/test/stix_transmission/test_transmission.py     422     52    88%
> -----------------------------------------------------------------------------------------------
> TOTAL                                                                        1277    181    86%
> ```

That looks great. Really appreciate to work on this. I'm merging the code.

**Review by mdazam1942 — APPROVED** [review]

File: stix_shifter_modules/infoblox/stix_translation/query_constructor.py [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+
+class QueryStringPatternTranslator:
+    comparator_lookup = {
+        'tideDbData': {
+            ComparisonExpressionOperators.And: "&",
+            ObservationOperators.And: '&',
+            ComparisonExpressionOperators.Or: "&",
+            ObservationOperators.Or: '&',
+            ComparisonComparators.Equal: "=",
+            ComparisonComparators.GreaterThan: "=",
```
Comment (delliott90) [inline stix_shifter_modules/infoblox/stix_translation/query_constructor.py:48]: @grimmjow8 I know this has already been merged, but we've been doing some work around the operators and noticed that you are assigning `=` to STIX operators that aren't equal. I don't fully understand how the query syntax for Infoblox works, but wouldn't this give a false result when using this in a query? It also looks like Infoblox only supports an AND operator, not OR. In cases where an operator is not supported, it should just be omitted from the comparator lookup.

Comment (grimmjow8) [reply stix_shifter_modules/infoblox/stix_translation/query_constructor.py:48]: Hey @delliott90! For the most part, it should give accurate results. Non-equal operators, like `>`,`>=`, etc are only supported by a few fields. Attempts to do that on non-capable fields should result in errors. For the fields that do support it, infoxblox api has particular field formats that are used. So for `imported` field, supported query parameters are `imported_from_date` and `imported_to_date`. Probably not the most intuitive out of the box but is still applicable.

In regards to `AND`/`OR`, we initially excluded it but it was recommended on this PR that a workaround could be to split it into multiple queries. We went with that since it seemed to provide accurate results.
